rfd/0205-improved-onprem-joining.md

```diff
@@ -0,0 +1,541 @@
+
+   Example UX:
+   ```
+   $ tctl bots add example --join-method=challenge
+   The bot token: challenge:04f0ceff1bd0589ba45c1832dfc8feaf
+   This token will expire in 59 minutes.
+   $ tbot start identity --token=challenge:04f0ceff1bd0589ba45c1832dfc8feaf
+   ```
+
+   On `tbot` startup, a keypair is transparently generated and exchanged with
+   Auth, after which the bot internally behaves as if flow 1 was used, and the
+   now-bound keypair perform its first full join.
+
+   From an end user's PoV, this process is nearly identical to traditional
+   `token` joining. While a joining secret does need to be copied to the bot
+   node, these secrets remain short-lived and one-time use.
+
+We expect most users to use Flow 2: it's much easier to provision new nodes and
+requires less back-and-forth between the admin's workstation and bot node. Flow
+1 is particularly ill-suited to Terraform use since keypairs would need to be
+pregenerated and copied to nodes, which is not ideal from a security PoV.
+
+#### Token Resource Example
+
+`challenge`-type tokens differ from other types in that they are intended to
+have no resource-level expiration (though that is allowed), are meant to have
+their spec modified over time by users or automation tools, and publish
+information about their current state in the immutable (to users) `status`
+field.
+
+``` yaml
+kind: token
+version: v2
+metadata:
+  name: my-join-token
+spec:
+  bot_name: example
+
+  join_method: challenge
+  challenge:
+
+    # `onboarding` parameters control initial join behavior
+    onboarding:
+      # If set, no joining secret is generated; the secret exchange ceremony is
+      # skipped and instance will directly prove its identity using its private
+      # key. It is an error for a public key to be associated with more than one
+      # token, and creation or update will fail if a public key is reused.
+      public_key: null
+
+      # If set, use an explicit initial joining secret; if both this and
+      # `public_key` are unset, a value will be generated server-side and
+      # stored in `.status.challenge.initial_join_secret`
+      initial_join_secret: ""
+
+      # Initial joining must take place before this timestamp. May be
+      # modified if bot has not yet joined.
+      expires: "2025-03-01T21:45:40.104524Z"
+
+    # Parameters to tune rejoining behavior when the regular bot identity has
+    # expired
+    rejoining:
+      # If true, `total_rejoins` is ignored and bots may rejoin indefinitely;
+      # must be opt-in.
+      unlimited: false
+
+      # Total number of allowed rejoins; this may be incremented to allow
+      # additional rejoins, even if a bot identity has already expired. May
+      # be decremented, but only by the current value of
+      # `.status.challenge.remaining_rejoins`.
+      total_rejoins: 10
+
+      # If set, rejoining is only valid before this timestamp; may be
+      # incremented to extend bot lifespan.
+      expires: ""
```

Comment:
Yeah, I'm thinking 3 different expiration times in the same resource is bound to get confusing.

Reply:
Yeah... this is the biggest sore spot I think. I think I might remove the resource-level expiration (or, well, make it an error to set one), and might drop this secondary expiration. Playing with some ideas to make the ergonomics here make sense, but it's confusing as is.

Reply:
As noted in another thread, I've renamed the duplicate fields. I'm not sure if that's ultimately a perfect solution but it's hopefully at least less confusing.

I do think (the field now known as) `must_rejoin_before` is optional. I think it's a reasonably sensible additional rejoining control admins might like to have available, but I think ultimately the design works alright with just a counter. If you think the end result is still confusing, we can remove this field.